pynicotine/gtkgui/userbrowse.py

```diff
@@ -688,34 +711,336 @@ def on_select_dir(self, selection):
+
+        if state & mods_alt:  # <GDK_MOD1_MASK>
+            prs_alt = True
+
+        if prs_ctrl == prs_shift == prs_alt is False:
+            prs_none = True  # <GDK_MOD2_MASK> (plain)
+
+        """ Key Bindings (this UserBrowses Page) """
+
+        # Ctrl+F: Focus Find TextEntry
+        keycodes, mods = parse_accelerator("<Primary>f")
+
+        if state & mods and keycode in keycodes and (not prs_shift and not prs_alt):
+            self.SearchEntry.grab_focus()
             return True
 
+        # Ctrl+G or F3: Find Next Match
+        keycodes_g, _mods_g = parse_accelerator("<Primary>g")
+        keycodes_f3, _mods_f3 = parse_accelerator("F3")
+
+        if (keycode in keycodes_g and prs_ctrl) or (keycode in keycodes_f3 and not prs_ctrl):
+            if (not prs_alt):  # Avoid any conflict
+                if (not prs_shift) and self.SearchEntry.get_text() != "":
+                    self.on_search()
+
+                elif prs_shift:
+                    # Shift+Ctrl+G or Shift+F3: Find Previous Match
+                    # ToDo: {self.search_position = self.search_position - 2}
+                    self.on_search()
+
+                else:
+                    self.SearchEntry.grab_focus()
+
+                return True
+
+        # Ctrl+R or F5: Refresh
+        keycodes_r, _mods_r = parse_accelerator("<Primary>r")
+        keycodes_f5, _mods_f5 = parse_accelerator("F5")
+
+        if (keycode in keycodes_r and prs_ctrl and not prs_shift) or (keycode in keycodes_f5 and prs_none):
+            if (not prs_alt):  # Avoid any conflict
+                self.on_refresh()
+                return True
+
+        # Ctrl+S: Save List
+        keycodes, mods = parse_accelerator("<Primary>s")
+
+        if state & mods and keycode in keycodes and (not prs_alt):
+            if (not prs_shift):
+                self.on_save()
+                return True
+
+            elif (prs_shift):
+                # Shift+Ctrl+S: Idea - Save List of all open tabs
+                return True
+
+        # Ctrl+U: Upload To User...
+        keycodes, mods = parse_accelerator("<Primary>u")
+
+        if state & mods and keycode in keycodes and (not prs_alt):
+            self.select_files()
+
+            if self.user == config.sections["server"]["login"]:
+                if self.num_selected_files >= 1:
+                    if (not prs_shift):
+                        self.on_upload_files()
+
+                    elif (prs_shift):
+                        self.on_upload_directory_to()
+
+                elif int(len(self.file_store)) >= 1 and self.num_selected_files <= 0 and (not prs_shift):
+                    self.on_upload_directory_to()
+
+                elif int(len(self.file_store)) <= 0 or (prs_shift):
+                    self.on_upload_directory_recursive_to()
+
+                return True
+
+            else:  # [user is not self]
+                return False
+
+        # Next key binding chain is self.Main
         return False
```

Comment:
This should accomplish what you want, if I understood your code correctly:
```
    def on_key_press_event_browse(self, *args):
        """ Key Bindings (this UserBrowses Page) """

        keyval, keycode, state, widget = get_key_press_event_args(*args)

        # Ctrl+F: Focus Find TextEntry
        keycodes, mods = parse_accelerator("<Primary>f")

        if state == mods and keycode in keycodes:
            self.SearchEntry.grab_focus()
            return True

        # Ctrl+G or F3: Find Next Match
        keycodes_g, mods_g = parse_accelerator("<Primary>g")
        keycodes_f3, mods_f3 = parse_accelerator("F3")

        if (state == mods_g and keycode in keycodes_g) or (state == mods_f3 and keycode in keycodes_f3):
            # Shift+Ctrl+G or Shift+F3: Find Previous Match
            # ToDo: {self.search_position = self.search_position - 2}
            self.on_search()
            return True

        # Shift+Ctrl+G or Shift+F3: Find Previous Match
        keycodes_g, mods_g = parse_accelerator("<Primary><Shift>g")
        keycodes_f3, mods_f3 = parse_accelerator("<Shift>F3")

        if (state == mods_g and keycode in keycodes_g) or (state == mods_f3 and keycode in keycodes_f3):
            # ToDo: {self.search_position = self.search_position - 2}
            self.on_search()
            return True

        # Ctrl+R or F5: Refresh
        keycodes_r, mods_r = parse_accelerator("<Primary>r")
        keycodes_f5, mods_f5 = parse_accelerator("F5")

        if (state == mods_r and keycode in keycodes_r) or (state == mods_f5 and keycode in keycodes_f5):
            self.on_refresh()
            return True

        # Ctrl+S: Save List
        keycodes, mods = parse_accelerator("<Primary>s")

        if state == mods and keycode in keycodes:
            self.on_save()
            return True

        # Shift+Ctrl+S: Idea - Save List of all open tabs
        keycodes, mods = parse_accelerator("<Primary><Shift>s")

        if state == mods and keycode in keycodes:
            return True

        # Ctrl+U: Upload To User...
        keycodes, mods = parse_accelerator("<Primary>u")

        if state == mods and keycode in keycodes:
            if self.user != config.sections["server"]["login"]:
                return False

            self.select_files()

            if self.num_selected_files >= 1:
                self.on_upload_files()

            elif len(self.file_store) >= 1 and self.num_selected_files <= 0:
                self.on_upload_directory_to()

            elif len(self.file_store) <= 0:
                self.on_upload_directory_recursive_to()

            return True

        # Shift+Ctrl+U: Upload Folder To User...
        keycodes, mods = parse_accelerator("<Primary><Shift>u")

        if state == mods and keycode in keycodes:
            if self.user != config.sections["server"]["login"]:
                return False

            self.select_files()

            if self.num_selected_files >= 1:
                self.on_upload_directory_to()
            else:
                self.on_upload_directory_recursive_to()

            return True

        # Next key binding chain is self.Main
        return False
```

Reply:
`if state == mods and keycode in keycodes` could probably be abstracted into a utility function? I could do this.

Reply:
It would look like this:
```
        # Ctrl+F: Focus Find TextEntry
        if accelerator_activated("<Primary>f", keycode, state):
            self.SearchEntry.grab_focus()
            return True
```

Reply:
> This should accomplish what you want

Your suggested code doesn't work, I think the bitwise & operator must be needed because the MOD2 (_no modifier keys pressed_) is always present, and this makes the raw variables different even when **_CONTROL_** is pressed and nothing else...

_str(state):_
`<flags GDK_CONTROL_MASK | GDK_MOD2_MASK of type Gdk.ModifierType>`

_str(mods):_
`<flags GDK_CONTROL_MASK of type Gdk.ModifierType>`

So unfortunately, `state` != `mods` (not equal: at least not on the two ThinkPad laptop systems I've tested it on).

> a utility function? I could do this.

This would be highly beneficial in the long run, as faffing about with these state and mod variables for every shortcut is not efficient, ideally any key combination could be tested on one line, and placed within any procedure definition, even if the key_event *args are not available (such as for Ctrl+MouseDblClick usage in on_folder_row_activated for example).

It is important that any extra pressed keys would cause your proposed utility function to return `False`, and only return `True` if the exact key combination is present with nothing more. In a perfect world, I imagine something like:

```
if get_key_press("<Primary>f"):
    self.SearchEntry.grab_focus()
```

Reply:
Your commit https://github.com/nicotine-plus/nicotine-plus/commit/a16bd4dc25d809241ac9ce3b63ab2abd518b6e1e will help greatly with these endeavors. The key bindings are going to be much more modular now, which is far better than giant on_key_press_event functions.